tests/store/tracking/test_file_store.py

```diff
@@ -2061,3 +2066,301 @@ def test_create_experiment_appends_to_artifact_local_path_file_uri_correctly(
+    )
+    dataset2 = Dataset(
+        name="name2",
+        digest="digest2",
+        source_type="st2",
+        source="source2",
+        schema="schema2",
+        profile="profile2",
+    )
+    dataset3 = Dataset(
+        name="name3",
+        digest="digest3",
+        source_type="st3",
+        source="source3",
+        schema="schema3",
+        profile="profile3",
+    )
+
+    tags1 = [InputTag(key="key1", value="value1"), InputTag(key="key2", value="value2")]
+    tags2 = [InputTag(key="key3", value="value3"), InputTag(key="key4", value="value4")]
+    tags3 = [InputTag(key="key5", value="value5"), InputTag(key="key6", value="value6")]
+
+    inputs_run1 = [DatasetInput(dataset1, tags1), DatasetInput(dataset2, tags1)]
+    inputs_run2 = [DatasetInput(dataset1, tags2), DatasetInput(dataset3, tags3)]
+    inputs_run3 = [DatasetInput(dataset2, tags3)]
+
+    store.log_inputs(run1.info.run_id, inputs_run1)
+    store.log_inputs(run2.info.run_id, inputs_run2)
+    store.log_inputs(run3.info.run_id, inputs_run3)
+
+    run1 = store.get_run(run1.info.run_id)
+    assert_dataset_inputs_equal(run1.inputs.dataset_inputs, inputs_run1)
+    run2 = store.get_run(run2.info.run_id)
+    assert_dataset_inputs_equal(run2.inputs.dataset_inputs, inputs_run2)
+    run3 = store.get_run(run3.info.run_id)
+    assert_dataset_inputs_equal(run3.inputs.dataset_inputs, inputs_run3)
+
+    search_results_1 = store.search_runs(
+        [exp_id], None, ViewType.ALL, max_results=4, order_by=["start_time ASC"]
+    )
+    run1 = search_results_1[0]
+    assert_dataset_inputs_equal(run1.inputs.dataset_inputs, inputs_run1)
+    run2 = search_results_1[2]
+    assert_dataset_inputs_equal(run2.inputs.dataset_inputs, inputs_run2)
+    run3 = search_results_1[1]
+    assert_dataset_inputs_equal(run3.inputs.dataset_inputs, inputs_run3)
+
+    search_results_2 = store.search_runs(
+        [exp_id], None, ViewType.ALL, max_results=4, order_by=["start_time DESC"]
+    )
+    run1 = search_results_2[2]
+    assert_dataset_inputs_equal(run1.inputs.dataset_inputs, inputs_run1)
+    run2 = search_results_2[0]
+    assert_dataset_inputs_equal(run2.inputs.dataset_inputs, inputs_run2)
+    run3 = search_results_2[1]
+    assert_dataset_inputs_equal(run3.inputs.dataset_inputs, inputs_run3)
+
+
+def test_log_input_multiple_times_does_not_overwrite_tags_or_dataset(store):
+    exp_id = store.create_experiment("dataset_no_overwrite")
+
+    run = store.create_run(
+        experiment_id=exp_id,
+        user_id="user",
+        start_time=0,
+        tags=[],
+        run_name=None,
+    )
+    dataset = Dataset(
+        name="name",
+        digest="digest",
+        source_type="st",
+        source="source",
+        schema="schema",
+        profile="profile",
+    )
+    tags = [InputTag(key="key1", value="value1"), InputTag(key="key2", value="value2")]
+    store.log_inputs(run.info.run_id, [DatasetInput(dataset, tags)])
+
+    for i in range(100):
```

Comment:
curious why we need do to 100 iterations of this vs just 1?

Reply:
Definitely don't need 100, but 1 also seems insufficient, e.g. perhaps the second write doesn't overwrite due to OBO instead of an intended design choice? I've changed to 3.